src/aulile/method.jl

```diff
@@ -0,0 +1,314 @@
+- `new_rules_symbol`: A symbol used to add new rules to the grammar as library learning.
+- `aux`: An `AuxFunction` that defines the evaluation metric and desired score.
+- `interpret`: An interpret function for the grammar.
+- `allow_evaluation_errors`: Whether to allow evaluation errors (such as in the grammar).
+- `max_iterations`: Maximum number of learning iterations to perform.
+- `max_depth`: Maximum depth for program enumeration.
+- `max_enumerations`: Maximum number of candidate programs to try per iteration.
+
+# Returns 
+- `SearchStats`: struct with the best program found, its score, number of iterations and enumerations.
+"""
+function aulile(
+        problem::Problem{<:AbstractVector{<:IOExample}},
+        iter_t::Type{<:ProgramIterator},
+        grammar::AbstractGrammar,
+        start_symbol::Symbol,
+        new_rules_symbol::Symbol,
+        aux::AuxFunction;
+        interpret::Function = default_interpreter,
+        allow_evaluation_errors::Bool = false,
+        max_iterations = 10000,
+        max_depth = 10,
+        max_enumerations = 100000
+)::SearchStats
+    iter = iter_t(grammar, start_symbol, max_depth = max_depth)
+    best_program = nothing
+    # Get initial distance of input and output
+    best_score = aux.initial_score(problem)
+    init_grammar_size = length(grammar.rules)
+    # Main loop
+    new_rules_decoding = Dict{Int, AbstractRuleNode}()
+    old_grammar_size = length(grammar.rules)
+    total_enumerations = 0
+    for i in 1:max_iterations
+        stats = synth_with_aux(problem, iter, grammar, aux,
+            new_rules_decoding, best_score,
+            interpret = interpret, allow_evaluation_errors = allow_evaluation_errors,
+            max_enumerations = max_enumerations)
+        total_enumerations += stats.enumerations
+        if stats.program isa Nothing
+            return SearchStats(best_program, stats.score, i, total_enumerations)
+        else
+            if best_score > 0
+                @assert stats.score < best_score
+            else
+                # In the case where the distance is optimal from the start
+                @assert stats.score <= best_score
+            end
+            best_program = stats.program
+            best_score = stats.score
+            if best_score <= aux.best_value
+                return SearchStats(stats.program, stats.score, i, total_enumerations)
+            else
+                program_expr = rulenode2expr(stats.program, grammar)
+                add_rule!(grammar, :($new_rules_symbol = $program_expr))
+                if length(grammar.rules) > old_grammar_size
+                    old_grammar_size = length(grammar.rules)
+                    new_rules_decoding[old_grammar_size] = deepcopy(stats.program)
+                end
+                iter = iter_t(grammar, start_symbol, max_depth = max_depth)
+            end
+        end
+    end
+    return SearchStats(best_program, best_score, max_iterations, total_enumerations)
```

Comment:
With `SearchStats` removed, we can just return a named tuple like this

```suggestion
    return (; best_program, best_score, max_iterations, total_enumerations)
```